dmvdata.py

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python3
+
+# SPDX-License-Identifier: GPL-2.0-only
+
+import json
+from json.decoder import JSONDecodeError
+import util
+import dmvutil
+from xcp import logger
+
+class DriverMultiVersionData:
+    def __init__(self, dmv_jsondata, hardware_info):
+        self.dmv_jsondata = dmv_jsondata
+        self.device_driver_map = hardware_info
+
+        dmvjson = None
+        dmvlist = None
+        try:
+            dmvjson = json.loads(dmv_jsondata)
+        except JSONDecodeError as e:
+            raise e
+        if not "drivers" in dmvjson:
+            raise RuntimeError("Invalid output from driver-tool")
+        dmvlist = dmvjson["drivers"]
+        self.drivers = dmvutil.parseDMVJsonData(dmvlist, hardware_info)
+        self.hw_present_drivers = dmvutil.getHardwarePresentDrivers(self.drivers)
+
+    def getDriversData(self):
+        return self.drivers
+
+    def getHardwarePresentDrivers(self):
+        return self.hw_present_drivers
+
+    def getHardwarePresentDriver(self, drvname):
+        return dmvutil.getHardwarePresentDriver(self.hw_present_drivers, drvname)
+
+    def queryDriversOrVariant(self, context):
+        return dmvutil.queryDriversOrVariant(context)
+
+    def getDriverVariantByName(self, drvname, oemtype):
+        return dmvutil.getDriverVariantByName(self.drivers, drvname, oemtype)
+
+    def sameDriverMultiVariantsSelected(self, variants):
+        return dmvutil.sameDriverMultiVariantsSelected(variants)
+
+    def chooseDefaultDriverVariants(self, drivers = None):
+        if not drivers:
+            drivers = self.hw_present_drivers
+        return dmvutil.chooseDefaultDriverVariants(drivers)
+
+    def selectSingleDriverVariant(self, driver_name, variant_name):
+        cmdstring = "driver-tool|-s|-n|%s|-v|%s" % (driver_name, variant_name)
+        rc, out = util.runCmd2(cmdstring.split('|'), with_stdout=True)
+        if rc != 0:
+            logger.log(out)
+            return False
+
+        rc, out = util.runCmd2(['rmmod', driver_name], with_stdout=True)
```

Comment:
I don't think this would work reliably.

If the driver is composed of multiple modules this will only rmmod one of them and then the modprobe could end of with a mix of modules.
If the driver is in use (e.g. because a RAID device was automatically assembled from the devices it provides), then rmmod will fail and the modprobe below will fail too.